[CodeGen][NewPM] Port RegAllocEvictionAdvisor analysis to NPM

optimisan opened the pull request:
Legacy pass used to provide the advisor, so this extracts that logic into a provider class used by both analysis passes.

All three (Default, Release, Development) legacy passes `*AdvisorAnalysis` are basically renamed to `*AdvisorProvider`, so the actual legacy wrapper passes are `*AdvisorAnalysisLegacy`.

There is only one NPM analysis `RegAllocEvictionAnalysis` that switches between the three providers in the `::run` method, to be cached by the NPM.

Also adds `RequireAnalysis<RegAllocEvictionAnalysis>` to the optimized target reg alloc codegen builder.

optimisan committed:
[CodeGen][NewPM] Port RegAllocEvictionAdvisor analysis to NPM

optimisan commented:
* **#117309** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117309?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/117309?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

paperchalice commented:
In general we should avoid module analysis results here. I browsed the relevant codes here, some machine learning based advisors need to load external models so they need `doInitialization`. Could we load and check models while building the pipeline?

cdevadas commented on llvm/lib/Passes/PassRegistry.def:
```diff
@@ -32,6 +32,7 @@ MODULE_ANALYSIS("no-op-module", NoOpModuleAnalysis())
 MODULE_ANALYSIS("pass-instrumentation", PassInstrumentationAnalysis(PIC))
 MODULE_ANALYSIS("profile-summary", ProfileSummaryAnalysis())
 MODULE_ANALYSIS("reg-usage", PhysicalRegisterUsageAnalysis())
+MODULE_ANALYSIS("regalloc-evict", RegAllocEvictionAdvisorAnalysis())
```
Alphabetical order.

mtrofin commented:
I'm trying to understand the goal here: is the intention to move codegen passes to NPM? I'd imagine there is a RFC for that, is this it? https://discourse.llvm.org/t/rfc-introducing-classes-for-the-codegen-driven-by-new-pass-manager/55877

I'm assuming you tried the change with the ML bits enabled? I.e. like the ml-opt-{dev|rel|devrel}-x86-64 bots do?

(I'd appreciate waiting for a LGTM from @boomanaiden154 or me, but please note next week is a generally a slow week in the US because the end of it is a bank holiday. I am out, but will pick this up the Monday after)

arsenm commented:
> I'm trying to understand the goal here: is the intention to move codegen passes to NPM? 

Yes, the actual pass doesn't matter. Every pass must be mechanically ported 

> I'm assuming you tried the change with the ML bits enabled? I.e. like the ml-opt-{dev|rel|devrel}-x86-64 bots do?

You cannot yet codeegn to the end. This is part of getting to a functional new pass manager invocation, there's nothing to test at this point

optimisan commented:
Yes
Can only test if this is not breaking the legacy pipeline (which I did) and if any standalone tests exist (for this analysis there aren't, will need to port greedy regalloc first)

mtrofin commented:
> > I'm trying to understand the goal here: is the intention to move codegen passes to NPM?
> 
> Yes, the actual pass doesn't matter. Every pass must be mechanically ported

OK, and to confirm, the RFC for that is the one I referenced?

> 
> > I'm assuming you tried the change with the ML bits enabled? I.e. like the ml-opt-{dev|rel|devrel}-x86-64 bots do?
> 
> You cannot yet codegen to the end. This is part of getting to a functional new pass manager invocation, there's nothing to test at this point

mtrofin commented:
What I'm trying to understand is how the design is expected to evolve. Is the current patch a crutch that enables then making RAGreedy NPM-enabled, and then we can drop the need for legacy support from the Regalloc advisor, remove that provider, remove LPM-isms like `doInitialization`, and basically end up with the same number of concepts we have today (i.e. before the current patch), but in NPM "terms"? 

(I read through the RFC but couldn't figure out what the approach to porting analyses was, sorry if I missed it)

arsenm commented:
> OK, and to confirm, the RFC for that is the one I referenced?

Yes, @paperchalice picked up those patches a while ago.

> What I'm trying to understand is how the design is expected to evolve. Is the current patch a crutch that enables then making RAGreedy NPM-enabled, and then we can drop the need for legacy support from the Regalloc advisor,

It's the same as porting IR passes to new PM. Both NewPM and legacy pass manager paths will need to coexist for the foreseeable future. So for now this is just additive, but someday the OldPM path will be deleted.

> basically end up with the same number of concepts we have today (i.e. before the current patch), but in NPM "terms"?

Yes

mtrofin commented:
Thanks. This really helps!

FWIW, it'd help if the RFC was updated to clarify these things (like 1 sentence like above brings a lot of clarity!), and if it were linked in these patches. I really like that we're moving off LPM, my concerns were the ones that got addressed above (i.e. progression/evolution of design). I should be able to review today.

Thanks!

mtrofin commented on llvm/include/llvm/CodeGen/RegAllocEvictionAdvisor.h:
```diff
@@ -191,13 +193,66 @@ class RegAllocEvictionAdvisorAnalysis : public ImmutablePass {
+/// This keeps the state for logging, and sets up and holds the provider.
+/// The legacy pass itself used to keep the logging state and provider,
+/// so this extraction helps the NPM analysis to reuse the logic.
+class RegAllocEvictionAdvisorProvider {
```
could you add a TODO here saying this is a point in time artifact for the period LPM and NPN coexist, after which this should be collapsed into the NPM analysis?

mtrofin commented on llvm/include/llvm/CodeGen/RegAllocEvictionAdvisor.h:
```diff
@@ -191,13 +193,66 @@ class RegAllocEvictionAdvisorAnalysis : public ImmutablePass {
+  const AdvisorMode Mode;
+};
+
+RegAllocEvictionAdvisorProvider *createReleaseModeAdvisorProvider();
```
do we need these `create` methods? Are they just an artifact that's not applicable anymore, coming from the LPM conversion?

mtrofin replied:
In fact, even more - I think you can drop the `doInitialization` from the providers and just pass the module in the ctor, since - iiuc - the `create...` functions are called in places you have a `Module`. In the LPM analyses' case, you can model the fields as `std::unique_ptr`s and dynamically initialze them there, wdyt?

optimisan replied:
I'll try that.

optimisan replied:
Can drop `doInitialization`, not sure how to work around declaring the `create..Advisor()`'s

The ML and Release mode providers are in `MLRegAlloc..cpp` so I wanted to expose only the "base" signature to create those providers in this default impl file `RegAllocEvictionAdvisor.cpp`

mtrofin replied:
Can't the respective analyses just instantiate the appropriate provider at their construction?

optimisan replied:
They do for the legacy analyses, but these are only used for the new (and only one) analysis.
Without the methods, the new PM analysis has to be in `MLRegAlloc..cpp` to keep the ML advisors local there, and also have to move the Default Provider out to `llvm::`.
Or - just separate out the ML advisor initialization in the new analysis to the `MLRegAlloc..cpp`?

mtrofin replied:
> Or - just separate out the ML advisor initialization in the new analysis to the `MLRegAlloc..cpp`?

Yes, each analysis would instantiate its provider as its implementation detail.

optimisan replied:
Removed them, separated out the ML initialization in the new PM analysis class to the ML implementation file.

mtrofin commented on llvm/include/llvm/Passes/CodeGenPassBuilder.h:
```diff
@@ -1061,9 +1062,10 @@ void CodeGenPassBuilder<Derived, TargetMachineT>::addMachineSSAOptimization(
     AddMachinePass &addPass, bool Optimized) const {
-  if (Optimized)
+  if (Optimized) {
+    addPass(RequireAnalysis<RegAllocEvictionAdvisorAnalysis>());
```
this doesn't work (well, but the CI already pointed that out).

optimisan replied:
Right, won't be necessary with a machine function analysis.

mtrofin commented on llvm/include/llvm/Passes/CodeGenPassBuilder.h:
```diff
@@ -1061,9 +1062,10 @@ void CodeGenPassBuilder<Derived, TargetMachineT>::addMachineSSAOptimization(
+    addPass(RequireAnalysis<RegAllocEvictionAdvisorAnalysis>());
     addPass(RAGreedyPass());
-  else
+  } else
```
style nit: if you added {} above, might as well do so for the `else`

mtrofin commented on llvm/include/llvm/CodeGen/RegAllocEvictionAdvisor.h:
```diff
@@ -191,13 +193,66 @@ class RegAllocEvictionAdvisorAnalysis : public ImmutablePass {
+
+  /// Set the analyses that the advisor needs to use as they might not be
+  /// available before the advisor is created.
+  virtual void setAnalyses(std::initializer_list<llvm::Any> AnalysisP) {}
```
does this need to be virtual? it always takes mbfi and loop info. might as well make it non-virtual and pass the explicit values

even better (connected with the `create...` comment) - why not just pass the analysis results in the ctor?

paperchalice commented:
I post a reply in RFC to summarize the work up to now: https://discourse.llvm.org/t/rfc-introducing-classes-for-the-codegen-driven-by-new-pass-manager/55877/26

optimisan commented:
> In general we should avoid module analysis results here. I browsed the relevant codes here, some machine learning based advisors need to load external models so they need `doInitialization`. Could we load and check models while building the pipeline?

Could you explain why to avoid module analysis results? 
Here iiuc we need to setup the Runner and Logger once, and then can provide the advisor. We can do that in the constructor of the NPM analysis and I'll change it to an MF analysis so this analysis owns the provider state (instead of being a module analysis).

paperchalice commented:
> Could you explain why to avoid module analysis results?

Just want to avoid outer analysis proxy here and let machine function analysis manager manage the advisors.

> Here iiuc we need to setup the Runner and Logger once, and then can provide the advisor which needs the Module context.

If I haven't missed anything, the context here is used only for error handling, we can handle them in pass builder.

optimisan commented:
Should I pass in the context from NewPMDriver into `CodeGenPassBuilder` for that?

optimisan committed:
Remove Provider::doInit, change to MF analysis

init is in the constructor. The new MF analysis lazily intiailizes the
provider. The (wrapped) provider is the analysis result.

optimisan commented:
Updated the NPM analysis to be a MachineFunction analysis.
- The analysis result is a pointer to the provider (pointer is owned by this analysis class, which lives throughout the pipeline run). So in NPM do `getResult<REAdvAnalysis>(MF).Provider.getAdvisor(MF, RAGreedy)`
- This analysis `::run` sets the new analysis to the same provider object.
- The provider is only invalidated if MBFI and MachineLoopInfo are invalidated since the provider's analyses are set by this `::run` (I could pass in the analyses in the `getAdvisor(MF, RAGreedy, MBFI, Loops)` method instead)

arsenm commented on llvm/lib/CodeGen/MLRegAllocEvictAdvisor.cpp:
```diff
@@ -405,14 +397,56 @@ class ReleaseModeEvictionAdvisorAnalysis final
-        MF, RA, Runner.get(),
-        getAnalysis<MachineBlockFrequencyInfoWrapperPass>().getMBFI(),
-        getAnalysis<MachineLoopInfoWrapperPass>().getLI());
+    assert((MBFI && Loops) &&
```
Don't need parentheses

arsenm commented on llvm/lib/CodeGen/MLRegAllocEvictAdvisor.cpp:
```diff
@@ -544,16 +565,60 @@ class DevelopmentModeEvictionAdvisorAnalysis final
-        MF, RA, Runner.get(),
-        getAnalysis<MachineBlockFrequencyInfoWrapperPass>().getMBFI(),
-        getAnalysis<MachineLoopInfoWrapperPass>().getLI(), Log.get());
+        MF, RA, Runner.get(), *MBFI, *Loops, Log.get());
```
Why isn't this std::move(Runner)?

optimisan replied:
Can change that in another patch

arsenm commented on llvm/lib/CodeGen/MLRegAllocEvictAdvisor.cpp:
```diff
@@ -535,7 +535,28 @@ class DevelopmentModeEvictionAdvisorAnalysis final
+    // contexts. At this point, it'd be an error
+    if (Log->currentContext() != MF.getName()) {
+      MF.getFunction().getContext().emitError(
+          "The training log context shouldn't have had changed.");
```
Don't use contractions in error messages (but this is probably preexisting)

arsenm commented on llvm/lib/CodeGen/MLRegAllocEvictAdvisor.cpp:
```diff
@@ -535,7 +535,28 @@ class DevelopmentModeEvictionAdvisorAnalysis final
+    // function, so we assume the last context belongs to this function. If
+    // this invariant ever changes, we can implement at that time switching
+    // contexts. At this point, it'd be an error
+    if (Log->currentContext() != MF.getName()) {
```
Relying on function names isn't great, there are anonymous functions...

optimisan committed:
Remove parenthesis

optimisan committed:
Add TODO to remove provider, remove paren

optimisan committed:
Remove createProvider*(), add new method in ML.cpp

optimisan committed:
Use the provider as analysis result

arsenm commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,110 @@ cl::opt<unsigned> EvictInterferenceCutoff(
+  if (Provider)
+    return;
+  if (Mode == RegAllocEvictionAdvisorAnalysisLegacy::AdvisorMode::Default)
+    Provider.reset(
+        new DefaultEvictionAdvisorProvider(/*NotAsRequested*/ false, Ctx));
+  else
+    initializeMLProvider(Mode, Ctx);
+  if (!Provider)
+    Provider.reset(
+        new DefaultEvictionAdvisorProvider(/*NotAsRequested*/ true, Ctx));
```
```suggestion
    Provider.reset(
        new DefaultEvictionAdvisorProvider(/*NotAsRequested=*/ false, Ctx));
  else
    initializeMLProvider(Mode, Ctx);
  if (!Provider)
    Provider.reset(
        new DefaultEvictionAdvisorProvider(/*NotAsRequested=*/ true, Ctx));
```

arsenm commented on llvm/lib/CodeGen/RegAllocGreedy.cpp:
```diff
@@ -2748,7 +2748,7 @@ bool RAGreedy::runOnMachineFunction(MachineFunction &mf) {
   ExtraInfo.emplace();
   EvictAdvisor =
-      getAnalysis<RegAllocEvictionAdvisorAnalysis>().getAdvisor(*MF, *this);
+      getAnalysis<RegAllocEvictionAdvisorAnalysisLegacy>().getProvider()->getAdvisor(*MF, *this);
```
Use a temporary variable here to avoid the ugly wrapping clang-format wants here

optimisan committed:
Remove setAnalyses(); pass analysis in getAdvisor

arsenm commented on llvm/include/llvm/CodeGen/RegAllocEvictionAdvisor.h:
```diff
@@ -164,40 +197,92 @@ class RegAllocEvictionAdvisor {
   /// Get an advisor for the given context (i.e. machine function, etc)
-  virtual std::unique_ptr<RegAllocEvictionAdvisor>
-  getAdvisor(const MachineFunction &MF, const RAGreedy &RA) = 0;
+  std::unique_ptr<RegAllocEvictionAdvisorProvider> &getProvider() {
+    return Provider;
+  }
```
```suggestion
  RegAllocEvictionAdvisorProvider &getProvider() {
    return *Provider;
  }
```

Returning the reference to the unique_ptr is weird

arsenm commented on llvm/include/llvm/CodeGen/RegAllocEvictionAdvisor.h:
```diff
@@ -164,40 +197,92 @@ class RegAllocEvictionAdvisor {
   AdvisorMode getAdvisorMode() const { return Mode; }
   virtual void logRewardIfNeeded(const MachineFunction &MF,
-                                 llvm::function_ref<float()> GetReward){};
+                                 llvm::function_ref<float()> GetReward) {};
```
Don't need llvm::

optimisan committed:
Apply suggestions

optimisan commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,102 @@ cl::opt<unsigned> EvictInterferenceCutoff(
+    RegAllocEvictionAdvisorAnalysisLegacy::AdvisorMode Mode, LLVMContext &Ctx) {
+  if (Provider)
+    return;
+  if (Mode == RegAllocEvictionAdvisorAnalysisLegacy::AdvisorMode::Default)
+    Provider.reset(
+        new DefaultEvictionAdvisorProvider(/*NotAsRequested=*/false, Ctx));
+  else
+    initializeMLProvider(Mode, Ctx);
+  if (!Provider)
+    Provider.reset(
+        new DefaultEvictionAdvisorProvider(/*NotAsRequested=*/true, Ctx));
```
Idea was to separate the ML implementation and take it out from this general file. This used to have two free functions like `createML..Provider()`

arsenm replied:
I don't see what's separated all. I'd rather just have duplicated code than whatever this is

optimisan replied:
Reverting to one switch case

optimisan committed:
Remove initializeMLProvider

arsenm commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,112 @@ cl::opt<unsigned> EvictInterferenceCutoff(
+    break;
+  }
+
+  if (!Provider)
```
Move to switch default

arsenm commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,112 @@ cl::opt<unsigned> EvictInterferenceCutoff(
+  case RegAllocEvictionAdvisorAnalysisLegacy::AdvisorMode::Default:
+    Provider.reset(
+        new DefaultEvictionAdvisorProvider(/*NotAsRequested=*/false, Ctx));
+    break;
```
return

optimisan committed:
address reviews, use #else

arsenm commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,113 @@ cl::opt<unsigned> EvictInterferenceCutoff(
+    Provider.reset(
+        new DefaultEvictionAdvisorProvider(/*NotAsRequested=*/true, Ctx));
+#endif
+    assert(Provider && "EvictionAdvisorProvider cannot be null");
```
Pretty sure this is a dead assert. new doesn't return null

arsenm commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,113 @@ cl::opt<unsigned> EvictInterferenceCutoff(
-  case RegAllocEvictionAdvisorAnalysis::AdvisorMode::Default:
-    Ret = new DefaultEvictionAdvisorAnalysis(/*NotAsRequested*/ false);
+  case RegAllocEvictionAdvisorAnalysisLegacy::AdvisorMode::Default:
+    return new DefaultEvictionAdvisorAnalysisLegacy(/*NotAsRequested*/ false);
```
```suggestion
    return new DefaultEvictionAdvisorAnalysisLegacy(/*NotAsRequested=*/ false);
```

arsenm commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,113 @@ cl::opt<unsigned> EvictInterferenceCutoff(
-  return new DefaultEvictionAdvisorAnalysis(/*NotAsRequested*/ true);
+
+  // release or development mode advisor may not be supported
+  if (!Ret)
```
Fully handle each case in the switch handling and don't have the break to conditional case down here

optimisan committed:
Apply AS, remove dead assert

arsenm commented on llvm/lib/CodeGen/RegAllocEvictionAdvisor.cpp:
```diff
@@ -59,59 +62,114 @@ cl::opt<unsigned> EvictInterferenceCutoff(
-    Ret = createDevelopmentModeAdvisor();
+    Ret = createDevelopmentModeAdvisorAnalysisLegacy();
+#else
+    Ret = new DefaultEvictionAdvisorAnalysisLegacy(/*NotAsRequested=*/true);
```
Directly return, no need for temporary Ret variable

optimisan committed:
simplify switch case with return

arsenm committed:
remove dead return

optimisan committed:
Simplify invalidate as the provider does not store analyses.
